### ISSUE TITLE
Use of CMAKE_<lang>_COMPILER_FORCED on host-builds flag breaks third-party libraries

### ISSUE THREAD

**Issue by palchak-google** [zephyrproject-rtos/zephyr#36558]
**Describe the bug**
When building for the "native_posix" target using the "host" toolchain, it's possible to use third-party libraries (e.g. unit test frameworks). Many such frameworks provide CMake build files, and some rely upon CMake's idiomatic, portable  `target_compile_features()` function to ensure the compiler supports all the necessary language features.

Zephyr's build system unconditionally sets the CMAKE_\<lang\>_COMPILER_FORCED global variables. This has the effect of inhibiting CMake's automatic discovery of compiler features. Without compiler feature discovery CMake generates an error at list time if `target_compile_features()` is used anywhere in the build.

**To Reproduce**
1. Add the following line to the end of `zephyr/samples/basic/minimal/CMakeLists.txt`:
`target_compile_features(app PRIVATE cxx_std_17)`
4. Attempt to build the samples/basic/minimal project with `BOARD=native_posix` and `ZEPHYR_TOOLCHAIN_VARIANT=host`:
```
$ ZEPHYR_TOOLCHAIN_VARIANT=host west build . --board=native_posix
```
5. See the following error:
```
CMake Error at CMakeLists.txt:21 (target_compile_features):
  target_compile_features no known features for CXX compiler

  "GNU"

  version 10.2.1.


-- Configuring incomplete, errors occurred!
```

**Expected behavior**
Build should complete normally and a valid `zephyr.elf` should be produced.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Host (gcc v10.2.1)
 

**Proposed Fix***

Zephyr should not set the CMAKE_\<lang\>_COMPILER_FORCED flags when the 'host' toolchain is used for the target. In such circumstances, the particulars of the host toolchain may not be known in advance (unlike what is typical with target-specific toolchains). Also the build may include third-party libraries that rely upon CMake's automatic discovery of compiler features.

The straightforward fix is as follows:

1) Remove the following lines from `zephyr/cmake/boilerplate.cmake`:
```
# Prevent CMake from testing the toolchain
set(CMAKE_C_COMPILER_FORCED   1)
set(CMAKE_CXX_COMPILER_FORCED 1)
```

2) Insert the same lines into `zephyr/cmake/target_toolchain.cmake` as follows:
```
if(NOT (COMPILER STREQUAL "host-gcc"))
  # Prevent CMake from testing the toolchain
  set(CMAKE_C_COMPILER_FORCED   1)
  set(CMAKE_CXX_COMPILER_FORCED 1)
  include(${TOOLCHAIN_ROOT}/cmake/toolchain/${ZEPHYR_TOOLCHAIN_VARIANT}/target.cmake)
endif()
```

**Comment by tejlmand** [comment]
The reason for this behavior is because Zephyr build infrastructure performs it's own toolchain testing regarding valid compiler flags.
On top of this, the testing result is cached. This significantly reduces CMake configure time, as the testing only needs to be carried out one time per each toolchain, and later CMake invocation in any project doesn't need to test the same toolchain again.

This is why CMake's own testing is disabled.

What we should probably do is to tell CMake about the known compile features when we are testing the compiler.
For example, if we know the toolchain supports `cxx_std_17` we should set:
```
set(CMAKE_CXX_COMPILE_FEATURES cxx_std_17)
```

This should probably be done in the compiler_flags.cmake for the compiler.

@palchak-google before posting a fix, I would like to test it out on your use-case, so could you provide a small snippet and a reference to the testing framework you are using. Thanks in advance.

**Comment by tejlmand** [comment]
> BTW, is this really a bug? It sounds like an enhancement.

I would say it depends on the direction from where you see this.
The fact that we have out own toolchain testing and thus breaks the way standard CMake works could be argued as a bug.
We shouldn't break built-in features of CMake (even though Zephyr itself doesn't use this feature).

But better support for downstream users can be argued as an enhancement.

But not a medium bug, so I changed to low.

But a fix is ready here: #39014

**Comment by tejlmand** [comment]
This is fixed.
